File: common/src/main/java/org/cosmic/ide/common/util/FileUtil.java
package org.cosmic.ide.common.util;

import java.io.BufferedWriter;
import java.io.File;
import java.io.FileFilter;
import java.io.FileWriter;
import java.io.IOException;
import java.io.InputStream;
import java.nio.file.FileVisitResult;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.nio.file.SimpleFileVisitor;
import java.nio.file.StandardCopyOption;
import java.nio.file.attribute.BasicFileAttributes;

public class FileUtil {

    private static String privateDataDirectory;

    public static void setDataDirectory(String directory) {
        privateDataDirectory = directory + "/";
    }

    public static boolean createDirectory(String path) {
        return new File(path).mkdir();
    }

    public static void writeFile(InputStream in, String path) throws IOException {
        var filePath = Paths.get(path).normalize();
        Files.copy(in, filePath, StandardCopyOption.REPLACE_EXISTING);
    }

    public static void writeFile(String path, String content) throws IOException {
        var file = new File(path);
        file.getParentFile().mkdirs();
        Files.write(file.toPath(), content.getBytes());
    }

    public static void writeFile(String path, byte[] content) throws IOException {
        var file = new File(path);
        file.getParentFile().mkdirs();
        Files.write(file.toPath(), content);
    }

    public static String readFile(File file) throws IOException {
        return new String(Files.readAllBytes(file.toPath()));
    }

    public static void deleteFile(String p) {
        try {
            var path = Paths.get(p);
            if (Files.isRegularFile(path)) {
                Files.delete(path);
                return;
            }

            Files.walkFileTree(
                    path,
                    new SimpleFileVisitor<Path>() {
                        @Override
                        public FileVisitResult visitFile(Path file, BasicFileAttributes attrs)
                                throws IOException {
                            Files.delete(file);
                            return FileVisitResult.CONTINUE;
                        }

                        @Override
                        public FileVisitResult postVisitDirectory(Path dir, IOException e)
                                throws IOException {
                            if (e != null) {
                                e.printStackTrace();
                            }
                            Files.delete(dir);
                            return FileVisitResult.CONTINUE;
                        }
                    });
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    public static String getDataDir() {
        return privateDataDirectory;
    }

    public static String getProjectsDir() {
        return getDataDir() + "projects/";
    }

    public static String getClasspathDir() {
        return getDataDir() + "classpath/";
    }

    public static boolean createOrExistsDir(final String dirPath) {
        return createOrExistsDir(getFileByPath(dirPath));
    }

    public static boolean createOrExistsDir(final File file) {
        return file != null && (file.exists() ? file.isDirectory() : file.mkdirs());
    }

    public static File getFileByPath(final String filePath) {
        return isSpace(filePath) ? null : new File(filePath);
    }

    private static boolean isSpace(final String s) {
        if (s == null) return true;
        for (char c : s.toCharArray()) {
            if (!Character.isWhitespace(c)) {
                return false;
            }
        }
        return true;
    }
}

File: common/src/main/kotlin/org/cosmic/ide/common/util/FileUtil.kt
package org.cosmic.ide.common.util

import java.io.File
import java.io.IOException
import java.io.InputStream
import java.nio.file.Files
import java.nio.file.Paths
import java.nio.file.StandardCopyOption

object FileUtil {

    private lateinit var privateDataDirectory: String

    @JvmStatic
    fun setDataDirectory(directory: String) {
        privateDataDirectory = directory + "/"
    }

    @JvmStatic
    fun createDirectory(path: String): Boolean {
        return File(path).mkdir()
    }

    @JvmStatic
    @Throws(IOException::class)
    fun writeFile(inp: InputStream, path: String) {
        val filePath = Paths.get(path).normalize()
        Files.copy(inp, filePath, StandardCopyOption.REPLACE_EXISTING)
    }

    @JvmStatic
    @Throws(IOException::class)
    fun writeFile(file: String, content: String) {
        File(file).writeText(content)
    }

    @JvmStatic
    @Throws(IOException::class)
    fun writeFile(path: String, content: ByteArray) {
        File(path).writeBytes(content)
    }

    @JvmStatic
    @Throws(IOException::class)
    fun readFile(file: File): String {
        return file.readText()
    }

    @JvmStatic
    fun deleteFile(path: String) {
        try {
            val file = File(path)
            if (file.isFile) {
                file.delete()
                return
            }

            file.deleteRecursively()
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }

    @JvmStatic
    fun getDataDir() = privateDataDirectory

    @JvmStatic
    fun getProjectsDir() = getDataDir() + "projects/"

    @JvmStatic
    fun getClasspathDir() = getDataDir() + "classpath/"

    @JvmStatic
    fun createOrExistsDir(dirPath: String) = createOrExistsDir(File(dirPath))

    @JvmStatic
    fun createOrExistsDir(file: File) = if (file.exists()) file.isDirectory else file.mkdirs()
}
